src/rdb/rdb_tx.c

```diff
@@ -417,20 +420,58 @@ rdb_tx_commit(struct rdb_tx *tx)
 	 */
 	if (!rdb_tx_is_critical(tx)) {
 		daos_size_t	scm_remaining = 0;
+		uint32_t	nchecks = 0;
 
+check_space:
 		rc = rdb_scm_left(tx->dt_db, &scm_remaining);
 		if (rc != 0) {
 			D_ERROR(DF_DB": failed to query free space\n",
 				DP_DB(tx->dt_db));
 			goto out_lock;
 		}
+		nchecks++;
 
 		if (scm_remaining < RDB_NOAPPEND_FREE_SPACE) {
-			D_DEBUG(DB_TRACE, DF_DB": nearly out of space, do not "
-			       "append! scm_left="DF_U64"\n", DP_DB(tx->dt_db),
-			       scm_remaining);
-			D_GOTO(out_lock, rc = -DER_NOSPACE);
+			uint64_t		idx = 0;
+			uint64_t		delta_nsec;
+
+			if (nchecks > 1) {
+				D_DEBUG(DB_TRACE, DF_DB": nearly out of space, do not append! "
+				       "scm_left="DF_U64"\n", DP_DB(tx->dt_db), scm_remaining);
+				D_GOTO(out_lock, rc = -DER_NOSPACE);
+			} else {
+				/* Throttle number of times compacting the log when low on space. */
+				delta_nsec = (d_hlc2nsec(d_hlc_get()) -
+					      d_hlc2nsec(tx->dt_db->d_nospc_ts));
+				if (delta_nsec < RDB_NOSPC_ERROR_INTVL_NSEC) {
+					D_DEBUG(DB_TRACE, DF_DB": nearly out of space, but too "
+						"early to trigger compaction\n", DP_DB(tx->dt_db));
+					goto check_space;
+				}
+			}
+
+			/* Trigger compaction of all applied entries since very low on space.
+			 * This may recover enough space to allow the operation to proceed.
+			 */
+			D_DEBUG(DB_TRACE, DF_DB": nearly out of space, compact log before retry! "
+				"scm_left="DF_U64"\n", DP_DB(tx->dt_db), scm_remaining);
+			rc = rdb_raft_trigger_compaction(tx->dt_db, true /* compact_all */, &idx);
+			if (rc != 0) {
+				D_WARN(DF_DB": failed to trigger compaction!\n", DP_DB(tx->dt_db));
+				D_GOTO(out_lock, rc = -DER_NOSPACE);
+			}
+			while ((idx != 0) && (tx->dt_db->d_lc_record.dlr_aggregated < idx)) {
+				sched_cond_wait(tx->dt_db->d_compacted_cv, tx->dt_db->d_raft_mutex);
+				D_DEBUG(DB_TRACE, DF_DB": compacted to "DF_U64", need "DF_U64"\n",
+					DP_DB(tx->dt_db), tx->dt_db->d_lc_record.dlr_aggregated,
+					idx);
+			}
+			tx->dt_db->d_nospc_ts = d_hlc_get();
+			goto check_space;
```

Comment:
[Question] Since we have released and reacquired `d_raft_mutex`, do we need to perform `rdb_tx_leader_check` again (e.g., in case the leader term no longer equals `tx->dt_term`)?

Reply:
In theory, we may lose leadership in term 10, and regain it in term 11 or higher. rdb_raft_append_apply will perform the append in this case. I admit, however, this may be just theoretical.